UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,695 @@
+            DEBUG ((DEBUG_INFO, "  %016lX\n", StartAddress));
+
+            PayloadBase->Entry = (EFI_PHYSICAL_ADDRESS)StartAddress;
+          }
+        }
+
+        if (AsciiStrnCmp (NodePtr->Name, "upl-params", AsciiStrLen ("upl-params")) == 0) {
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "addr-width", &TempLen);
+          if (TempLen > 0) {
+            Data32 = (UINT32 *)(PropertyPtr->Data);
+            DEBUG ((DEBUG_INFO, "\n         Property(00000000)  address_width"));
+            DEBUG ((DEBUG_INFO, "  %X", Fdt32ToCpu (*Data32)));
+            SizeOfMemorySpace = (UINT8)Fdt32ToCpu (*Data32);
+            BuildCpuHob (SizeOfMemorySpace, PcdGet8 (SizeOfIoSpace));
+          }
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "pci-enum-done", &TempLen);
+          if (TempLen > 0) {
+            Data16      = (UINT16 *)(PropertyPtr->Data);
+            PciEnumDone = (UINT8)Fdt16ToCpu (*Data16);
+            DEBUG ((DEBUG_INFO, "  Found PciEnumDone (%08X)\n", PciEnumDone));
+          }
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "boot-mode", &TempLen);
+          if (TempLen > 0) {
+            TempStr = (CHAR8 *)(PropertyPtr->Data);
+            if (AsciiStrCmp (TempStr, "normal") == 0) {
+              BootMode = BOOT_WITH_FULL_CONFIGURATION;
+            } else if (AsciiStrCmp (TempStr, "fast") == 0) {
+              BootMode = BOOT_WITH_MINIMAL_CONFIGURATION;
+            } else if (AsciiStrCmp (TempStr, "full") == 0) {
+              BootMode = BOOT_WITH_FULL_CONFIGURATION_PLUS_DIAGNOSTICS;
+            } else if (AsciiStrCmp (TempStr, "default") == 0) {
+              BootMode = BOOT_WITH_DEFAULT_SETTINGS;
+            } else if (AsciiStrCmp (TempStr, "s4") == 0) {
+              BootMode = BOOT_ON_S4_RESUME;
+            } else if (AsciiStrCmp (TempStr, "s3") == 0) {
+              BootMode = BOOT_ON_S3_RESUME;
+            }
+          }
+        }
+      }
+    }
+    // Optional
+    else if (AsciiStrnCmp (NodePtr->Name, "pci-rb", AsciiStrLen ("pci-rb")) == 0) {
+      DEBUG ((DEBUG_INFO, "  Found pci-rb node (%08X)", Node));
+
+      HobDataSize = sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES) + RootBridgeCount *sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGE);
+      //
+      // Create PCI Root Bridge Info Hob.
+      //
+      if (PciRootBridgeInfo == NULL) {
+        PciRootBridgeInfo = BuildGuidHob (&gUniversalPayloadPciRootBridgeInfoGuid, HobDataSize);
+        ASSERT (PciRootBridgeInfo != NULL);
+        if (PciRootBridgeInfo == NULL) {
+          break;
+        }
+
+        ZeroMem (PciRootBridgeInfo, HobDataSize);
+        PciRootBridgeInfo->Header.Length    = sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES);
+        PciRootBridgeInfo->Header.Revision  = UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES_REVISION;
+        PciRootBridgeInfo->Count            = RootBridgeCount;
+        PciRootBridgeInfo->ResourceAssigned = (BOOLEAN) PciEnumDone;

```

Comment:
Why don't we keep single source of truth? PcdPciDisableBusEnumeration align to this PCD?

Reply:
PcdPciDisableBusEnumeration is not generic way for configuration cross boot loaders, so UPL should take FDT root node->pci-enum-done as generic input and internally UPL will use PciRootBridgeInfo->ResourceAssigned to update PcdPciDisableBusEnumeration.

Reply:
I see the code which aligns this PCD. I guess then it should be fine.